File: packages/rook_sdk_apple_health/lib/src/data/mapper/sleep_summary_mappers.dart
import 'package:rook_sdk_apple_health/src/data/extension/datetime_extensions.dart';
import 'package:rook_sdk_apple_health/src/data/mapper/complex_extensions.dart';
import 'package:rook_sdk_apple_health/src/data/mapper/granular_data_mappers.dart';
import 'package:rook_sdk_apple_health/src/data/proto/protos.pb.dart';
import 'package:rook_sdk_apple_health/src/domain/util/check_non_default.dart';
import 'package:rook_sdk_core/rook_sdk_core.dart';

extension SleepSummaryMappers on SleepSummaryProto {
  SleepSummary toDomain() {
    return SleepSummary(
      dateTime: DateTimeExtensions.parseLocal(dateTime),
      sourceOfData: sourceOfData,
      wasTheUserUnderPhysicalActivity: wasTheUserUnderPhysicalActivity,
      sleepStartDateTime: DateTimeExtensions.parseLocal(sleepStartDateTime),
      sleepEndDateTime: DateTimeExtensions.parseLocal(sleepEndDateTime),
      sleepDate: DateTimeExtensions.parseLocal(sleepDate),
      sleepDurationSeconds: sleepDurationSeconds.checkNonDefault(),
      timeInBedSeconds: timeInBedSeconds.checkNonDefault(),
      lightSleepDurationSeconds: lightSleepDurationSeconds.checkNonDefault(),
      remSleepDurationSeconds: remSleepDurationSeconds.checkNonDefault(),
      deepSleepDurationSeconds: deepSleepDurationSeconds.checkNonDefault(),
      timeToFallAsleepSeconds: timeToFallAsleepSeconds.checkNonDefault(),
      timeAwakeDuringSleepSeconds:
          timeAwakeDuringSleepSeconds.checkNonDefault(),
      sleepQualityRating1To5Score:
          sleepQualityRating1To5Score.checkNonDefault(),
      sleepEfficiency1To100Score: sleepEfficiency1To100Score.checkNonDefault(),
      sleepGoalSeconds: sleepGoalSeconds.checkNonDefault(),
      sleepContinuity1To5Score: sleepContinuity1To5Score.checkNonDefault(),
      sleepContinuity1To5Rating: sleepContinuity1To5Rating.checkNonDefault(),
      hrMaximumBpm: hrMaximumBpm.checkNonDefault(),
      hrMinimumBpm: hrMinimumBpm.checkNonDefault(),
      hrAvgBpm: hrAvgBpm.checkNonDefault(),
      hrRestingBpm: hrRestingBpm.checkNonDefault(),
      hrBasalBpm: hrBasalBpm.checkNonDefault(),
      hrGranularData: hrGranularData.map((element) {
        return element.toDomain();
      }).toList(),
      hrvAvgRmssd: hrvAvgRmssd.checkNonDefault(),
      hrvAvgSdnn: hrvAvgSdnn.checkNonDefault(),
      hrvSdnnGranularData: hrvSdnnGranularData.map((element) {
        return element.toDomain();
      }).toList(),
      hrvRmssdGranularData: hrvRmssdGranularData.map((element) {
        return element.toDomain();
      }).toList(),
      temperatureMinimum: temperatureMinimum
          .map((element) {
            return element.toDomain();
          })
          .toList()
          .firstOrNull,
      temperatureAvg: temperatureAvg
          .map((element) {
            return element.toDomain();
          })
          .toList()
          .firstOrNull,
      temperatureMaximum: temperatureMaximum
          .map((element) {
            return element.toDomain();
          })
          .toList()
          .firstOrNull,
      temperatureGranularData: temperatureGranularData.map((element) {
        return element.toDomain();
      }).toList(),
      temperatureDelta: temperatureDelta
          .map((element) {
            return element.toDomain();
          })
          .toList()
          .firstOrNull,
      breathsMinimumPerMin: breathsMinimumPerMin.checkNonDefault(),
      breathsAvgPerMin: breathsAvgPerMin.checkNonDefault(),
      breathsMaximumPerMin: breathsMaximumPerMin.checkNonDefault(),
      breathingGranularData: breathingGranularData.map((element) {
        return element.toDomain();
      }).toList(),
      snoringEventsCount: snoringEventsCount.checkNonDefault(),
      snoringDurationTotalSeconds: snoringDurationTotalSeconds,
      snoringGranularData: snoringGranularData.map((element) {
        return element.toDomain();
      }).toList(),
      saturationAvgPercentage: saturationAvgPercentage.checkNonDefault(),
      saturationMinimumPercentage:
          saturationMinimumPercentage.checkNonDefault(),
      saturationMaximumPercentage:
          saturationMaximumPercentage.checkNonDefault(),
      saturationGranularData: saturationGranularData.map((element) {
        return element.toDomain();
      }).toList(),
    );
  }
}

File: packages/rook_sdk_apple_health/ios/Classes/Mapper/SleepSummaryMappers.swift
//
//  SleepSummaryMappers.swift
//  rook_sdk_apple_health
//
//  Created by Daniel Nolasco on 28/10/25.
//

import Foundation
import RookSDK

extension RookSleepSummary {
    func toProto() -> SleepSummaryProto {
        return SleepSummaryProto.with {
            $0.dateTime = datetime.toProtoString()
            $0.sourceOfData = sourceOfData
            $0.wasTheUserUnderPhysicalActivity = false
            $0.sleepStartDateTime = sleepStartDatetime.toProtoString()
            $0.sleepEndDateTime = sleepEndDatetime.toProtoString()
            $0.sleepDate = sleepDate.toProtoString()
            $0.sleepDurationSeconds = sleepDurationSeconds.toInt32()
            $0.timeInBedSeconds = timeInBedSeconds.toInt32()
            $0.lightSleepDurationSeconds = lightSleepDurationSeconds.toInt32()
            $0.remSleepDurationSeconds = remSleepDurationSeconds.toInt32()
            $0.deepSleepDurationSeconds = deepSleepDurationSeconds.toInt32()
            $0.timeToFallAsleepSeconds = timeToFallAsleepSeconds.toInt32()
            $0.timeAwakeDuringSleepSeconds = timeAwakeDuringSleepSeconds.toInt32()
            $0.sleepQualityRating1To5Score = sleepQualityRating1_5_Score.toInt32()
            $0.sleepEfficiency1To100Score = sleepEfficiency1_100_Score.toInt32()
            $0.sleepGoalSeconds = sleepGoalSeconds.toInt32()
            $0.sleepContinuity1To5Score = sleepContinuity1_5_Score.toInt32()
            $0.sleepContinuity1To5Rating = sleepContinuity1_5_Rating.toInt32()
            $0.hrMaximumBpm = hrMaxBPM.toInt32()
            $0.hrMinimumBpm = hrMinimumBPM.toInt32()
            $0.hrAvgBpm = hrAvgBPM.toInt32()
            $0.hrRestingBpm = hrRestingBPM.toInt32()
            $0.hrBasalBpm = hrBasalBPM.toInt32()
            $0.hrGranularData = hrGranularDataBPM?.map { $0.toGranularProto() } ?? []
            $0.hrvAvgRmssd = hrvAvgRmssdNumber.toDouble()
            $0.hrvAvgSdnn = hrvAvgSdnnNumber.toDouble()
            $0.hrvSdnnGranularData = hrvSdnnGranularData?.map { $0.toGranularProto() } ?? []
            $0.hrvRmssdGranularData = hrvRmssdGranularData?.map { $0.toGranularProto() } ?? []
            $0.temperatureMinimum = temperatureMinimumCelsius?.map { $0.toObjectProto() } ?? []
            $0.temperatureAvg = temperatureAvgCelsius?.map { $0.toObjectProto() } ?? []
            $0.temperatureMaximum = temperatureMaxCelsius?.map { $0.toObjectProto() } ?? []
            $0.temperatureGranularData = temperatureGranularDataCelsius?.map { $0.toGranularProto() } ?? []
            $0.temperatureDelta = temperatureDeltaCelsius?.map { $0.toObjectProto() } ?? []
            $0.breathsMinimumPerMin = breathsMinimumPerMin.toDouble()
            $0.breathsAvgPerMin = breathsAvgPerMin.toDouble()
            $0.breathsMaximumPerMin = breathsMaxPerMin.toDouble()
            $0.breathingGranularData = breathingGranularDataBreathsPerMin?.map { $0.toGranularProto() } ?? []
            $0.snoringEventsCount = snoringEventsCountNumber.toInt32()
            $0.snoringDurationTotalSeconds = snoringDurationTotalSeconds.toInt32()
            $0.snoringGranularData = snoringGranularDataSnores?.map { $0.toGranularProto() } ?? []
            $0.saturationAvgPercentage = saturationAvgPercentage.toDouble()
            $0.saturationMinimumPercentage = saturationMinPercentage.toDouble()
            $0.saturationMaximumPercentage = saturationMaxPercentage.toDouble()
            $0.saturationGranularData = saturationGranularDataPercentage?.map { $0.toGranularProto() } ?? []
        }
    }
}
